File: easy/self_describing_numbers.c
#include <stdio.h>

int main(int argc, char *argv[])
{
    FILE *fp;
    char line[12];

    fp = fopen(*++argv, "r");
    while (fgets(line, 12, fp) != 0) {
	int i = 0, j = 1, a[10] = {0}, b[10] = {0};
	while (line [i] != '\n') {
	    a[i] = line[i] - 48;
	    b[a[i++]]++;
	}
	while (i-- > 0)
	    if (a[i] != b[i])
		j = 0;
	printf("%d\n", j);
    }
    return 0;
}

File: easy/self_describing_numbers.rb
File.open(ARGV[0]).each_line do |line|
  a, n = Array.new, line.to_i
  line.chomp.each_char do |i|
    a << 0
  end
  m, selfn = n, true
  while m > 0 do
    v = m % 10
    if v < a.length
      a[v] += 1
    else
      selfn = false
      break
    end
    m /= 10
  end
  if selfn
    r = 0
    (0...a.length).each do |i|
      r = r*10 + a[i]
    end
    selfn = false if n != r
  end
  puts selfn ? 1 : 0
end
